src/NHibernate/Cache/StandardQueryCache.cs

```diff
@@ -106,60 +143,250 @@ public IList Get(QueryKey key, ICacheAssembler[] returnTypes, bool isNaturalKeyL
+			for (var i = 0; i < keys.Length; i++)
+			{
+				var result = results[i];
+				if (queryParameters[i].NaturalKeyLookup && result.Count == 0)
+					continue;
+
+				cached[i] = true;
+				cachedKeys.Add(keys[i]);
+				cachedResults.Add(GetCacheableResult(returnTypes[i], session, result, ts));
+			}
+
+			_batchableCache.PutMany(cachedKeys.ToArray(), cachedResults.ToArray());
+
+			return cached;
+		}
+
+		/// <inheritdoc />
+		public IList[] GetMany(
+			QueryKey[] keys,
+			QueryParameters[] queryParameters,
+			ICacheAssembler[][] returnTypes,
+			ISet<string>[] spaces,
+			ISessionImplementor session)
+		{
+			var results = new IList[keys.Length];
+			if (_batchableReadOnlyCache == null)
+			{
+				for (var i = 0; i < keys.Length; i++)
+				{
+					results[i] = Get(keys[i], queryParameters[i], returnTypes[i], spaces[i], session);
+				}
+				return results;
+			}
+
+			if (Log.IsDebugEnabled())
+				Log.Debug("checking cached query results in region: '{0}'; {1}", _regionName, StringHelper.CollectionToString(keys));
+
+			var cacheables = _batchableReadOnlyCache.GetMany(keys).Cast<IList>().ToArray();
+
+			var spacesToCheck = new List<ISet<string>>();
+			var checkedSpacesIndexes = new HashSet<int>();
+			var checkedSpacesTimestamp = new List<long>();
+			for (var i = 0; i < keys.Length; i++)
+			{
+				var cacheable = cacheables[i];
+				if (cacheable == null)
+				{
+					Log.Debug("query results were not found in cache: {0}", keys[i]);
+					continue;
+				}
+
+				var querySpaces = spaces[i];
+				if (queryParameters[i].NaturalKeyLookup || querySpaces.Count == 0)
+					continue;
+
+				spacesToCheck.Add(querySpaces);
+				checkedSpacesIndexes.Add(i);
+				// The timestamp is the first element of the cache result.
+				checkedSpacesTimestamp.Add((long) cacheable[0]);
+				if (Log.IsDebugEnabled())
+					Log.Debug("Checking query spaces for up-to-dateness [{0}]", StringHelper.CollectionToString(querySpaces));
+			}
+
+			var upToDates = spacesToCheck.Count > 0
+				? _updateTimestampsCache.AreUpToDate(spacesToCheck.ToArray(), checkedSpacesTimestamp.ToArray())
+				: Array.Empty<bool>();
+
+			var upToDatesIndex = 0;
+			var persistenceContext = session.PersistenceContext;
+			var defaultReadOnlyOrig = persistenceContext.DefaultReadOnly;
+			for (var i = 0; i < keys.Length; i++)
+			{
+				var cacheable = cacheables[i];
+				if (cacheable == null)
+					continue;
+
+				var key = keys[i];
+				if (checkedSpacesIndexes.Contains(i) && !upToDates[upToDatesIndex++])
+				{
+					Log.Debug("cached query results were not up to date for: {0}", key);
+					continue;
+				}
+
+				var queryParams = queryParameters[i];
+				if (queryParams.IsReadOnlyInitialized)
+					persistenceContext.DefaultReadOnly = queryParams.ReadOnly;
+				else
+					queryParams.ReadOnly = persistenceContext.DefaultReadOnly;
+
+				try
+				{
+					results[i] = GetResultFromCacheable(key, returnTypes[i], queryParams.NaturalKeyLookup, session, cacheable);
```

Comment:
Here we are in a loop on each query for handling one query result per iteration. But `GetResultFromCacheable` may also trigger cache gets, in case the query result contains some `ManyToOneType` to `Assemble` by example.
In such case, an immediate load is triggered. It will cause entity cache gets, which will be batched through the `batch-size` mechanism, since #1633: the result is looped for calling `BeforeAssemble` before being looped a second time for actually assembling. `BeforeAssemble` does schedule batching of the entities, with the lazy load batching mechanism.
So there is still some batching of those entity cache gets too, but it is not well suited to the query cache get case (including single query gets):
- It may get other entities unrelated to the queries being loaded, potentially with an inadequate readonly status on the persistence context. This trouble is not new, it could already occurs previously on single query get, through batch loading from database when the entity was not cached (either by user choice or due to a cache miss for whatever reason). But it has somewhat be worsen by #1633, since it may now occur without a cache miss due to entity cache get batching using the `batch-size` mechanism.
- It is restricted to `batch-size` entities, as set on the entity mapping. In the query case, it would be better to batch the gets of all the entities matching the query.
- It is segregated by entity persister, while again in the query case, it would  be better to batch them all together.
- It is segregated by query, while in the multiple query get case, it could be batched across queries sharing the same readonly status.

For handling all that, a dedicated cache batching mechanism would be required, allowing to batch the entity cache gets of the queries together without batching with them other pending gets. It should also take care of the readonly status of the persistence context.  The last point (the only one specific to multi-query get) would moreover require to handle the `BeforeAssemble` calls in another dedicated loop on queries in order to batch them together before doing another loop for actually assembling all the queries.

That may be an improvement for later. I have decided not to do it here to avoid expanding a bit too much the scope of this PR.